### ISSUE TITLE
Weird error

### ISSUE THREAD

**Issue by ColinChargy** [bankscrap/bankscrap-ing#8]
Hi,
I'm on Debian and I've run :
```
sudo apt install ruby-dev
sudo gem install bankscrap-ing
```
and finally :
`bankscrap balance ING --credentials=dni:<MY_CLIENT_NUMBER> password:<MY_SECRET_CODE> birthday:<MY_BIRTHDAY>`

And it return the following log :

```
/var/lib/gems/2.3.0/gems/mechanize-2.7.6/lib/mechanize/http/agent.rb:329:in `fetch': 400 => Net::HTTPBadRequest for https://ing.ingdirect.es/genoma_login/rest/session -- unhandled response (Mechanize::ResponseCodeError)
        from /var/lib/gems/2.3.0/gems/mechanize-2.7.6/lib/mechanize.rb:562:in `request_with_entity'
        from /var/lib/gems/2.3.0/gems/mechanize-2.7.6/lib/mechanize.rb:511:in `post'
        from /var/lib/gems/2.3.0/gems/bankscrap-2.1.1/lib/bankscrap/bank.rb:65:in `post'
        from /var/lib/gems/2.3.0/gems/bankscrap-ing-2.0.4/lib/bankscrap/ing/bank.rb:108:in `request_pinpad_positions'
        from /var/lib/gems/2.3.0/gems/bankscrap-ing-2.0.4/lib/bankscrap/ing/bank.rb:97:in `login'
        from /var/lib/gems/2.3.0/gems/bankscrap-2.1.1/lib/bankscrap/bank.rb:29:in `initialize'
        from /var/lib/gems/2.3.0/gems/bankscrap-ing-2.0.4/lib/bankscrap/ing/bank.rb:24:in `initialize'
        from /var/lib/gems/2.3.0/gems/bankscrap-2.1.1/lib/bankscrap/cli.rb:118:in `new'
        from /var/lib/gems/2.3.0/gems/bankscrap-2.1.1/lib/bankscrap/cli.rb:118:in `initialize_client_for'
        from /var/lib/gems/2.3.0/gems/bankscrap-2.1.1/lib/bankscrap/cli.rb:24:in `balance'
        from /var/lib/gems/2.3.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
        from /var/lib/gems/2.3.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
        from /var/lib/gems/2.3.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
        from /var/lib/gems/2.3.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
        from /var/lib/gems/2.3.0/gems/bankscrap-2.1.1/bin/bankscrap:5:in `<top (required)>'
        from /usr/local/bin/bankscrap:22:in `load'
        from /usr/local/bin/bankscrap:22:in `<main>'
```

Did I do anything wrong ? How can I have more info on what's wrong ?
Is the DNI my client number ? Is the secret code the usual secret code I've use on the ING webite or mobile app ?
Regards,
Colin Chargy

**Comment by raulmarcosl** [comment]
Hi Colin,

To have more info you can always add `--debug` to your request, but in this case, it only shows the headers used. We should improve how we handle errors.

Regarding your problem, `dni` refers to DNI, the Spanish identity card. I develop this gem, and as I'm based in Spain I connect to ING using that, so if you are using ING from another country probably you have to use your ID number (supposing the web and the login are similar).

The secret code yes, is the six-number digit you use to log in (ING requests three random positions of that number, but the gem needs the whole number).

Please let me know if you have any problem. And if it finally works from another country, I'm interested to know more!

**Comment by ColinChargy** [comment]
Hi @raulmarcosl ,
Thanks for your feedback!
Here is the log with debug :
```
D, [2019-04-09T09:31:50.419320 #3517] DEBUG -- : query: "{\"loginDocument\":{\"documentType\":0,\"document\":\"3901002\"},\"birthday\":\"22/09/1990\",\"companyDocument\":null,\"device\":\"desktop\"}"
I, [2019-04-09T09:31:50.423277 #3517]  INFO -- : Net::HTTP::Post: /genoma_login/rest/session
D, [2019-04-09T09:31:50.425867 #3517] DEBUG -- : request-header: accept-encoding => gzip,deflate,identity
D, [2019-04-09T09:31:50.425993 #3517] DEBUG -- : request-header: accept => application/json, text/javascript, */*; q=0.01
D, [2019-04-09T09:31:50.426086 #3517] DEBUG -- : request-header: user-agent => Mozilla/5.0 (Linux; Android 4.2.1; en-us; Nexus 4 Build/JOP40D) AppleWebKit/535.19 (KHTML, like Gecko) Chrome/18.0.1025.166 Mobile Safari/535.19
D, [2019-04-09T09:31:50.426168 #3517] DEBUG -- : request-header: accept-charset => ISO-8859-1,utf-8;q=0.7,*;q=0.7
D, [2019-04-09T09:31:50.426225 #3517] DEBUG -- : request-header: accept-language => en-us,en;q=0.5
D, [2019-04-09T09:31:50.426281 #3517] DEBUG -- : request-header: host => ing.ingdirect.es
D, [2019-04-09T09:31:50.426341 #3517] DEBUG -- : request-header: content-type => application/json; charset=utf-8
D, [2019-04-09T09:31:50.426395 #3517] DEBUG -- : request-header: content-length => 123
I, [2019-04-09T09:31:50.854687 #3517]  INFO -- : status: Net::HTTPBadRequest 1.1 400 Bad Request
D, [2019-04-09T09:31:50.854819 #3517] DEBUG -- : response-header: date => Tue, 09 Apr 2019 07:23:16 GMT
D, [2019-04-09T09:31:50.854843 #3517] DEBUG -- : response-header: server => Apache
D, [2019-04-09T09:31:50.854856 #3517] DEBUG -- : response-header: strict-transport-security => max-age=631138519; includeSubDomains
D, [2019-04-09T09:31:50.854889 #3517] DEBUG -- : response-header: pragma => no-cache, no-cache
D, [2019-04-09T09:31:50.854908 #3517] DEBUG -- : response-header: expires => Thu, 01 Jan 1970 00:00:00 GMT, Thu, 01 Jan 1970 00:00:00 GMT
D, [2019-04-09T09:31:50.854919 #3517] DEBUG -- : response-header: cache-control => no-cache, no-store, no-cache, no-store, max-age=0
D, [2019-04-09T09:31:50.854930 #3517] DEBUG -- : response-header: content-language => en-US
D, [2019-04-09T09:31:50.854940 #3517] DEBUG -- : response-header: x-xss-protection => 1; mode=block
D, [2019-04-09T09:31:50.854951 #3517] DEBUG -- : response-header: x-content-type-options => nosniff
D, [2019-04-09T09:31:50.854961 #3517] DEBUG -- : response-header: set-cookie => genoma-session-id=a60b4ce9-09be-4469-8c17-1b56de7217b5; Path=/genoma_login/; Secure; HttpOnly
D, [2019-04-09T09:31:50.854972 #3517] DEBUG -- : response-header: vary => User-Agent,Accept-Encoding
D, [2019-04-09T09:31:50.854982 #3517] DEBUG -- : response-header: content-encoding => gzip
D, [2019-04-09T09:31:50.854994 #3517] DEBUG -- : response-header: connection => close
D, [2019-04-09T09:31:50.855010 #3517] DEBUG -- : response-header: transfer-encoding => chunked
D, [2019-04-09T09:31:50.855026 #3517] DEBUG -- : response-header: content-type => application/json;charset=UTF-8
D, [2019-04-09T09:31:50.855198 #3517] DEBUG -- : Read 97 bytes (97 total)
D, [2019-04-09T09:31:50.855288 #3517] DEBUG -- : Read 10 bytes (107 total)
D, [2019-04-09T09:31:50.855635 #3517] DEBUG -- : gzip response
D, [2019-04-09T09:31:50.858944 #3517] DEBUG -- : saved cookie: genoma-session-id=a60b4ce9-09be-4469-8c17-1b56de7217b5
/var/lib/gems/2.3.0/gems/mechanize-2.7.6/lib/mechanize/http/agent.rb:329:in `fetch': 400 => Net::HTTPBadRequest for https://ing.ingdirect.es/genoma_login/rest/session -- unhandled response (Mechanize::ResponseCodeError)
        from /var/lib/gems/2.3.0/gems/mechanize-2.7.6/lib/mechanize.rb:562:in `request_with_entity'
        from /var/lib/gems/2.3.0/gems/mechanize-2.7.6/lib/mechanize.rb:511:in `post'
        from /var/lib/gems/2.3.0/gems/bankscrap-2.1.1/lib/bankscrap/bank.rb:65:in `post'
        from /var/lib/gems/2.3.0/gems/bankscrap-ing-2.0.4/lib/bankscrap/ing/bank.rb:108:in `request_pinpad_positions'
        from /var/lib/gems/2.3.0/gems/bankscrap-ing-2.0.4/lib/bankscrap/ing/bank.rb:97:in `login'
        from /var/lib/gems/2.3.0/gems/bankscrap-2.1.1/lib/bankscrap/bank.rb:29:in `initialize'
        from /var/lib/gems/2.3.0/gems/bankscrap-ing-2.0.4/lib/bankscrap/ing/bank.rb:24:in `initialize'
        from /var/lib/gems/2.3.0/gems/bankscrap-2.1.1/lib/bankscrap/cli.rb:118:in `new'
        from /var/lib/gems/2.3.0/gems/bankscrap-2.1.1/lib/bankscrap/cli.rb:118:in `initialize_client_for'
        from /var/lib/gems/2.3.0/gems/bankscrap-2.1.1/lib/bankscrap/cli.rb:24:in `balance'
        from /var/lib/gems/2.3.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
        from /var/lib/gems/2.3.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
        from /var/lib/gems/2.3.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
        from /var/lib/gems/2.3.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
        from /var/lib/gems/2.3.0/gems/bankscrap-2.1.1/bin/bankscrap:5:in `<top (required)>'
        from /usr/local/bin/bankscrap:22:in `load'
        from /usr/local/bin/bankscrap:22:in `<main>'
```

I just tried with [https://ing.ingdirect.es/pfm/#login/credentials](https://ing.ingdirect.es/pfm/#login/credentials) and my french ING client number is not recognized by the Spanish ING website. I can't find the REST API for ING in France. How did you find the Spanish one ? I see that now ING has an open data API [https://developer.ing.com/](https://developer.ing.com/), maybe we should use this API so all ING customer could access their account with this script.

Regards,
Colin

**Comment by raulmarcosl** [comment]
As I can see, ING France uses another different site and API (the login is using https://m.ing.fr/secure/api-v1/login/cif`), so this gem doesn't work for France. 

To find the Spanish version of their API we did reverse engineering of their mobile app, it's not public. Their open data API looks promising, it didn't exist when I did this, but as it happens with other banks, it requires signing up, so we believe it's great also to have a permissionless way to access your own data.
